kitsune/wiki/views.py

```diff
@@ -471,106 +471,144 @@ def edit_document(request, document_slug, revision_id=None):
         url = reverse("wiki.translate", locale=request.LANGUAGE_CODE, args=[document_slug])
         return HttpResponseRedirect(url)
 
-    can_edit_needs_change = doc.allows(user, "edit_needs_change")
-    can_archive = doc.allows(user, "archive")
-
     # If this document has a parent, then the edit is handled by the
     # translate view. Pass it on.
     if doc.parent:
         return translate(request, doc.parent.slug, revision_id)
+
+    if not doc.allows(user, "create_revision"):
+        raise PermissionDenied
+
     if revision_id:
         rev = get_object_or_404(Revision, pk=revision_id, document=doc)
     else:
         rev = doc.current_revision or doc.revisions.order_by("-created", "-id")[0]
 
-    disclose_description = bool(request.GET.get("opendescription"))
-    doc_form = rev_form = None
-    if doc.allows(user, "create_revision"):
+    if request.method == "GET":
```

Comment:
Can't we replace this block with something like:

```python
    rev_form = RevisionForm(
        request.POST or None, instance=rev, initial={"based_on": rev.id, "comment": ""}
    )
    if rev_form.is_valid():
        rev_form.instance.doc = doc
        _document_lock_clear(doc.id, user.username)
        _save_rev_and_notify(rev_form, user, doc, base_rev=rev)
        if "notify-future-changes" in request.POST:
            EditDocumentEvent.notify(request.user, doc)
        return HttpResponseRedirect(reverse("wiki.document_revisions", args=[document_slug]))

    show_revision_warning = _show_revision_warning(doc, rev)
    locked, locked_by = _document_lock(doc.id, user.username)

    return render(
        request,
        "wiki/edit.html",
        {
            "revision_form": rev_form,
            "document": doc,
            "show_revision_warning": show_revision_warning,
            "locked": locked,
            "locked_by": locked_by,
        },
    )
```

Reply:
Your version breaks tests - because it is behaving differently than the original code. In your code, on a POST request, `instance` and `initial` aren't discarded when the form is re-instantiated with POST data. So when `is_valid()` is called, the POST data is validated against the model `instance` constraints and the `initial` data.